lib/maths/CAdaptiveBucketing.cc

```diff
@@ -400,16 +403,68 @@ bool CAdaptiveBucketing::knots(core_t::TTime time,
                 break;
 
             case CSplineTypes::E_Periodic:
-                values[0] = (values[0] + values.back()) / 2.0;
-                variances[0] = (variances[0] + variances.back()) / 2.0;
-                knots.push_back(m_Endpoints[n]);
-                values.push_back(values[0]);
-                variances.push_back(variances[0]);
+                // We search for the value in the last and next period which
+                // are adjacent. Note that values need not be the same at the
+                // start and end of the period because the gradient can vary,
+                // but we expect them to be continuous.
+                for (std::size_t j = n - 1; j > 0; --j) {
```

Comment:
Have you compared performance after this change?

Reply:
This only happens very infrequently (when we interpolate which is once per period) so typically once per day, week, etc. As a result these functions are never bottlenecks. For this specific code, n is small and typically this loop exits after the first iteration (only carries on for sparse data). The unit tests didn't noticeably slowdown, so I think this is probably safe from that perspective.